### PR TITLE
Added capability to quantize a model while exporting through ONNX.

### PR DESCRIPTION
**Pull request by mfuntowicz**
Add quantization support as part of our collaboration with ONNX team.

Quantization is available through the new method `quantize` which takes the path to the initial ONNX model and operate the conversion.

From a CLI point of view, adding `--quantize` allows to seamlessly export the quantized model.

Signed-off-by: Morgan Funtowicz <morgan@huggingface.co>

### CODE REVIEW

**Comment by mfuntowicz** [comment]
cc: @tianleiwu @yufenglee

File: src/transformers/convert_graph_to_onnx.py [hunk 1]
```diff
@@ -37,9 +38,41 @@ def __init__(self):
...
 
 
+def generate_identified_filename(filename: Path, identifier: str) -> Path:
+    """
+    Append an string-identifier at the end (before the extension,  if any) to the provided filepath.
+    Can be used to
+    :param filename:
+    :param identifier:
+    :return:
+    """
```
Comment (LysandreJik) [inline src/transformers/convert_graph_to_onnx.py:52]: Maybe at one point this doc could be updated to fit the google style that we use everywhere else!

**Review by LysandreJik — APPROVED** [review]
Very cool. Int 8 quantization is impressive!

File: src/transformers/convert_graph_to_onnx.py (hunk 1 above)
Comment (mfuntowicz) [reply src/transformers/convert_graph_to_onnx.py:52]: Yes ofc, will update the docstrings, sorry for this ...

File: src/transformers/convert_graph_to_onnx.py [hunk 2]
```diff
@@ -30,16 +30,66 @@ class OnnxConverterArgumentParser(ArgumentParser):
...
+    Check onnxruntime is installed and if the installed version match is recent enough.
+    Raises:
+        ImportError: If onnxruntime is not installed or too old version is found
+    """
+    try:
+        import onnxruntime
+
+        # We require 1.4.0 minimum
+        if int(onnxruntime.__version__.split(".")[1]) < 4:
```
Comment (tianleiwu) [inline src/transformers/convert_graph_to_onnx.py:78]: For example, version 2.0 cannot pass this check. How about using packaging.version to compare version.

File: src/transformers/convert_graph_to_onnx.py [hunk 3]
```diff
@@ -214,25 +320,60 @@ def convert(
         convert_tensorflow(nlp, opset, output)
 
 
-def verify(path: str):
+def quantize(onnx_model_path: Path) -> Path:
```
Comment (yufenglee) [inline src/transformers/convert_graph_to_onnx.py:323]: Thanks Morgan for integrating onnx quantization! This is the overall process of onnx quantization. For transformer-based models, it gets better performance by optimizing with onnxruntime_tools.optimizer before quantizing. Below is an example to optimize Bert-base with the tool:

# optimize transformer-based models with onnxruntime-tools
from onnxruntime_tools import optimizer
from onnxruntime_tools.transformers.onnx_model_bert import BertOptimizationOptions

# disable embedding layer norm optimization for better model size reduction
opt_options = BertOptimizationOptions('bert')
opt_options.enable_embed_layer_norm = False

opt_model = optimizer.optimize_model(
    'bert.onnx',
    'bert', 
    num_heads=12,
    hidden_size=768,
    optimization_options=opt_options)
opt_model.save_model_to_file('bert.opt.onnx')

onnx_opt_model = onnx.load('bert.opt.onnx')
quantized_onnx_model = quantize(onnx_opt_model,
                                    quantization_mode=QuantizationMode.IntegerOps,
                                    symmetric_weight=True,
                                    force_fusions=True)

onnx.save(quantized_onnx_model, 'bert.opt.quant.onnx')

File: src/transformers/convert_graph_to_onnx.py [hunk 4]
```diff
@@ -30,16 +30,66 @@ class OnnxConverterArgumentParser(ArgumentParser):
     def __init__(self):
         super(OnnxConverterArgumentParser, self).__init__("ONNX Converter")
```
Comment (sshleifer) [inline src/transformers/convert_graph_to_onnx.py:31]: (nit) you can just write `super().__init__`

File: src/transformers/convert_graph_to_onnx.py [hunk 5]
```diff
@@ -30,16 +30,66 @@ class OnnxConverterArgumentParser(ArgumentParser):
...
+        self.add_argument(
+            "--framework", type=str, choices=["pt", "tf"], help="Framework for loading the model",
+        )
         self.add_argument("--opset", type=int, default=11, help="ONNX opset to use")
-        self.add_argument("--check-loading", action="store_true", help="Check ONNX is able to load the model")
-        self.add_argument("--use-external-format", action="store_true", help="Allow exporting model >= than 2Gb")
+        self.add_argument(
+            "--check-loading", action="store_true", help="Check ONNX is able to load the model",
+        )
+        self.add_argument(
```
Comment (sshleifer) [inline src/transformers/convert_graph_to_onnx.py:47]: These flags feel like they should all default to true.
Why would I use this if I didn't want to quantize?

**Review by sshleifer — APPROVED** [review]

File: src/transformers/convert_graph_to_onnx.py [hunk 6]
```diff
@@ -214,25 +320,60 @@ def convert(
...
+    try:
+        ensure_onnxruntime_installed()
+        import onnx
+        from onnxruntime import __version__ as ort_version
+        from onnxruntime.quantization import quantize, QuantizationMode
+
+        print(f"Found ONNX: {onnx.__version__}")
+        print(f"Found ONNXRuntime: {ort_version}")
+
+        onnx_model = onnx.load(onnx_model_path.as_posix())
```
Comment (yufenglee) [inline src/transformers/convert_graph_to_onnx.py:341]: Thanks Morgan for integrating onnx quantization! For transformer-based models, it gets better performance to  optimize the model before quantizing:

```
from onnxruntime import SessionOptions, InferenceSession

# optimize the model and save it
opt_model_path = generate_identified_filename(onnx_model_path, "-opt")
sess_option = SessionOptions()
sess_option.optimized_model_filepath = opt_model_path
sess = InferenceSession(onnx_model_path, sess_option)

# load and quantize optimized model
onnx_model = onnx.load(opt_model_path.as_posix())
quantized_model = quantize(
            model=onnx_model, quantization_mode=QuantizationMode.IntegerOps, force_fusions=True, symmetric_weight=True,
)

# Save model
print(f"Storing quantized model at {quantized_model_path}")
onnx.save(quantized_model, quantized_model_path.as_posix())

```

File: src/transformers/convert_graph_to_onnx.py (hunk 2 above)
Comment (mfuntowicz) [reply src/transformers/convert_graph_to_onnx.py:78]: Thanks for pointing out @tianleiwu, that's one of package (_packaging.version_) I'm not super used to play with so didn't investigate much more. Will have a look 👍

File: src/transformers/convert_graph_to_onnx.py (hunk 6 above)
Comment (mfuntowicz) [reply src/transformers/convert_graph_to_onnx.py:341]: @yufenglee I've a dev branch which will bring the optimisations you mentioned into the export script. I decided to keep them in separate PR for easier reviewing 😃. 

Should be in before the end of the week 👍

Comment (mfuntowicz) [reply src/transformers/convert_graph_to_onnx.py:350]: How actually it's exactly the same way of optimizing as I was following the content of the notebook you included upstream [in ONNX runtime](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/quantization/notebooks/Bert-GLUE_OnnxRuntime_quantization.ipynb).

Which way would be better @yufenglee ? 

```python
# optimize transformer-based models with onnxruntime-tools
from onnxruntime_tools import optimizer
from onnxruntime_tools.transformers.onnx_model_bert import BertOptimizationOptions

# disable embedding layer norm optimization for better model size reduction
opt_options = BertOptimizationOptions('bert')
opt_options.enable_embed_layer_norm = False

opt_model = optimizer.optimize_model(
    'bert.onnx',
    'bert', 
    num_heads=12,
    hidden_size=768,
    optimization_options=opt_options)
opt_model.save_model_to_file('bert.opt.onnx')
```

**Review by tianleiwu — APPROVED** [review]
LGTM

File: src/transformers/convert_graph_to_onnx.py (hunk 6 above)
Comment (yufenglee) [reply src/transformers/convert_graph_to_onnx.py:351]: @mfuntowicz, onnxruntime_tools is more flexible, for example, you can enable/disable certain optimization.

**Review by yufenglee — APPROVED** [review]

**Pull request merged**